Refactor NGP

ProstoSanja opened the pull request.

ProstoSanja commented:
Now it looks like this
![image](https://cloud.githubusercontent.com/assets/26527928/24166943/b5eeda7a-0e7d-11e7-9093-608b682e0e44.png)
do not merge yet

ProstoSanja commented:
Currently pushing this version here
![image](https://cloud.githubusercontent.com/assets/26527928/24169372/16815f86-0e86-11e7-8f95-511b56f51f70.png)

need you opinion

ProstoSanja commented:
Whats better:
this 
![image](https://cloud.githubusercontent.com/assets/26527928/24187006/b3b5b33a-0ee3-11e7-81a8-58c113e58212.png)
or this 
![image](https://cloud.githubusercontent.com/assets/26527928/24187032/d408fa34-0ee3-11e7-8efb-562587013c37.png)

Chronophylos commented:
2nd one.

ProstoSanja commented:
Currently push this version
![image](https://cloud.githubusercontent.com/assets/26527928/24189655/865ec326-0eee-11e7-8827-c31b3f1d7f61.png)

Chronophylos commented:
Looks great!

ProstoSanja commented:
Fixed lua and js errors and optimized ux, added timer to ui, because it was at backend anyway, and the gnp didnt disapprear after it ended
![image](https://cloud.githubusercontent.com/assets/26527928/24193379/bc9fd9d2-0efa-11e7-98ca-59055ad70c06.png)

Chronophylos commented on game/scripts/vscripts/components/boss/ngp.lua:
```diff
@@ -26,15 +26,14 @@ function NGP:PlayerVote (eventSourceIndex, args)
   local id = eventSourceIndex.id
   local option = eventSourceIndex.option
   local team = ShortTeamName(PlayerResource:GetTeam(playerID))
-  local item = NGP.activeItems[tonumber(id)]
```
You don't need to do that, in fact you shouldn't. `item` is a reference to `NGP.activeItems[tonumber(id)]`. `item` and `NGP.activeItems[tonumber(id)]` are effectively the same. But `item` is shorter and easier to read.
I didn't mean you should replace it.

ProstoSanja replied:
okay

ProstoSanja commented:
Fixed a code to work with frontend, 
before voting
![image](https://cloud.githubusercontent.com/assets/26527928/24194630/6ed43f86-0eff-11e7-98e2-a2865505eb57.png)
after voting
![image](https://cloud.githubusercontent.com/assets/26527928/24194492/fc7436e4-0efe-11e7-9b27-583ad662238f.png)

Chronophylos commented:
I think after voting should look like before voting but the selected button is highlighted

ProstoSanja commented:
okay, maybe, any other opinions?

ProstoSanja commented:
Ready for merge/discussion

ProstoSanja commented:
Actually found error at line :61
```
[   VScript       ]: scripts/vscripts/components\boss\ngp.lua:61: attempt to index a nil value
[   VScript       ]: stack traceback:
[   VScript       ]: 	scripts/vscripts/libraries\timers.lua:137: in function '__newindex'
[   VScript       ]: 	scripts/vscripts/components\boss\ngp.lua:61: in function <scripts/vscripts/components\boss\ngp.lua:59>
[   VScript       ]: 	[C]: in function 'xpcall'
[   VScript       ]: 	scripts/vscripts/libraries\timers.lua:136: in function <scripts/vscripts/libraries\timers.lua:94>
[   VScript       ]: 
```

Chronophylos commented on game/scripts/vscripts/components/boss/ngp.lua:
```diff
@@ -43,8 +43,9 @@ function NGP:GiveItemToTeam (item, team)
+  item.finished = false
   NGP.itemIndex = NGP.itemIndex + 1
-
+  DebugPrint('XD ')
```
Remove this

ProstoSanja replied:
okay, but what with that error? it is about `ngpItems`

ProstoSanja commented:
cant seem to understand, it is 50 50, either no crash and gives item, either this mistake

ProstoSanja commented:
Finally ready for merge

ProstoSanja commented:
Do not merge yet

ProstoSanja commented:
@chrisinajar I need help with beforehand end of voting.
Lua scripts work absolutely fine and give all iteams, if the team has finished voting,
but even though i receive item, the panels for first two ngps (withh ids 0 and 1) only go away after timer.
And exactly two first ngp items, no matter from what boss.
Any ideas?

ProstoSanja commented:
Done with ngp, looks like this when you havent woted
![image](https://cloud.githubusercontent.com/assets/26527928/24242426/61c8ac34-0fc0-11e7-8435-6cd56a4b2e1c.png)
And like this when you have
![image](https://cloud.githubusercontent.com/assets/26527928/24242441/6a0c0da0-0fc0-11e7-87b6-0ffee9a4a8e4.png)

Also adds icons of people who voted

Bugtested and ready to be merged

Trildar commented:
There's a typo in line 12 of `need_greed_pass/panel.xml`. Not your typo, but might as well fix it while you're changing this stuff. The references to that class in `need_greed_pass.css` also have the typo. I'm guessing they were auto-completed or something. Also check anywhere else you think might refer to that class.

ProstoSanja commented:
Ok, I'll look through it

darklordabc commented:
How long are those popups there for?

Trildar commented:
For line 11 in `need_greed_pass.js`, you can change it to use the defined constant `DOTATeam_t.DOTA_TEAM_GOODGUYS`. Possibly rewrite the line altogether, particularly since using the constant would make the line quite long.

ProstoSanja commented:
They are there for initial time of ngp(60 seconds, didn't touch that), or until everyone has voted

darklordabc commented:
Ah good, I was going to suggest that they dissapear after x amount of time so they dont stack up too much.

ProstoSanja commented:
Will not rewite ja variables, because this is a useless suggestion. We are not planning on implementing another teams, so no need in that. Also I'll have to change backend then again, and I don't want to rewrite same thing, which is already working.

darklordabc commented:
I agree not worth rewrite, but in general you typically use the names DOTA_TEAM_BADGUYS, etc, because the number value may be changed by value some day if they add some weird new team for some event, and it would break all the code that depends on a certain team being a certain number. But yeah, not worth rewrite.

ProstoSanja commented:
Well, that's good point. I'll look it through

Trildar commented:
Well, just to clarify, I didn't mean to change the way the net table was handled or anything. I meant change the ternary operator into a full if statement cause that might be easier to read.

Trildar commented:
Correct me if I'm wrong, but shouldn't the `RemoveNeedGreedPass` function also clean up the corresponding id in `existingPanels`? Also, does the `NGP.itemIndex` just keep increasing? Guess pretty much all of that would fall under further refactoring NGP though.

ProstoSanja commented:
No it shouldn't: active panels -> panels which were already handled, and have gone to archive (activated panels is better name)
Itemindex is the I'd of ngp to throw, each ngp has a unique one

chrisinajar commented:
The buttons need to remain visible, the point of it staying there is so you can change your vote.

chrisinajar commented on game/scripts/vscripts/components/boss/ngp.lua:
```diff
@@ -5,8 +5,13 @@ if NGP == nil then
+NGP.itemIndex = 2
 NGP.activeItems = {}
+NGP.activeTimers = {}
+totalgoodplayers = 0
```
don't use global variables

ProstoSanja replied:
do you mean like remove NGP prefix?

ProstoSanja commented:
Can you?

ProstoSanja commented:
Before this refactor panel immediately disappeared.
Okay, I'll look into it

ProstoSanja commented:
After fifth player it just dismisses itself, same as before behavior

chrisinajar commented:
It didn't do that before (but should), i mean i want to see what it looks like for the 5th player, not after. between 4th and 5th.

Trildar commented:
If players are allowed to change their vote, doesn't it make more sense to just have a fixed time?

ProstoSanja commented:
@chrisinajar are you sure they should be able to dismiss?

ProstoSanja commented:
ready for check and merge

ProstoSanja commented:
@chrisinajar i am done!

ProstoSanja commented:
actually i have a thought, why do we five upgrade core lvl1 via ngp at t2 boss? 
Shouldnt it be like ngp lvl2 + combiner, and lvl1 to all? it would also solve a bug(tested)
Also, tested with guys, no more team mismatch problem occurs

Trildar commented:
Level 2 Upgrade Core upgrades items to level 3. T1 boss isn't supposed to drop any Upgrade Cores.

ProstoSanja commented:
we are talking about t2 boss, which imo should give lvl1 upcore to everyone in team.

ProstoSanja commented:
Will finish tomorrow

ProstoSanja commented:
i made progress with our bug, it is in lua, exactly third panel does not apply `finished` to itself

ProstoSanja commented:
Okay found the problem, it updates local variable in lua, but for some reason doesnt update nettable exactly at id 4

ProstoSanja commented:
More info, code works as intendet, and local `NGP.activeItems` works perfectly, but for some unknown reason exactly during vote for the third ngp, it doesnt update nettable. I tried creating timers and so on, delaying it and it still doesnt update exactly on third item, as if it skips `setTableItem` for no apparent reason, because code after this one `FinishVoting` continues running, i dont have any clue, why this might happen...
Maybe we should just add another gameevent to force close panels in addition to existing nettable?

ProstoSanja commented:
posting my latest code:
changes
1) polished everything a little
2) map changes: added `version` variable to t1 bosses with values 1 and 2. for reflex core and farming core one
3) split t1 boss to t1a and t1b
4) added fallback that if no `version at spawnpoint or boss doesnt have a version`, to use random boss from tier
5) fixed boss drops at levels (thanks balance guys)

Trildar commented on game/scripts/vscripts/components/boss/ngp.lua:
```diff
@@ -17,50 +22,115 @@ function NGP:Init ()
-  DebugPrintTable(ngpItems)
   item.id = NGP.itemIndex
+  item.finished = false
   NGP.itemIndex = NGP.itemIndex + 1
```
I know the code was like this when you started working on it, but I'm wondering if all these references to `NGP` inside the functions shouldn't be `self` instead. Guess they probably both mean the same thing in this case.

ProstoSanja replied:
Not good in lua, so i use what i know that works

Trildar commented on game/scripts/vscripts/components/boss/ngp.lua:
```diff
@@ -5,8 +5,13 @@ if NGP == nil then
+local totalgoodplayers = 0
+local totalbadplayers = 0
+local ngpItemsGood = {}
+local ngpItemsBad = {}
```
I'm not sure this is actually important, but I feel like all these file-scope local variables ought to be properties on the NGP class like `itemIndex` and `activeItems` are.

ProstoSanja replied:
These items are local things, they can be re-retrieved in every other class using standart methods, they dont need to be global

Trildar commented on game/scripts/vscripts/components/boss/ngp.lua:
```diff
@@ -17,50 +22,115 @@ function NGP:Init ()
-  Timers:CreateTimer(60, function ()
-    NGP:FinishVoting(NGP.activeItems[item.id])
-  end)
+function getGlobalTable(team) 
```
`getGlobalTable` and the other net table functions here should be within the NGP class. i.e. declare as `function NGP:getGlobalTable`. Otherwise the functions will be in global scope.

You would then call them with `self:getGlobalTable`, etc.

Trildar commented on game/scripts/vscripts/components/boss/ngp.lua:
```diff
@@ -17,50 +22,115 @@ function NGP:Init ()
+      if player:GetTeam() == 3 then
+        totalbadplayers = totalbadplayers + 1
+      elseif player:GetTeam() == 2 then
+        totalgoodplayers = totalgoodplayers + 1
```
Getting the total number of players on a team can be done using the built-in `PlayerResource:GetPlayerCountForTeam` function.

ProstoSanja commented:
ill fix map, because my was utdated, and then this will be finally ready

ProstoSanja commented:
@chrisinajar GOD YES, IT IS DONE

ProstoSanja commented:
all map changes are described above, and it can be safely replaced with mine, because i specially updated it with latest your version before commit

Chronophylos commented:
You still have conflicting files.

ProstoSanja commented:
Dont merge how you did in morning

ProstoSanja commented:
the map should be replaced, it is a modified version from couple of hours ago, and the secind file is very easy, even from first sight

ProstoSanja commented:
Like i need @chrisinajar to merge this, cuz it is unavailable to resolve conflict for me
![image](https://cloud.githubusercontent.com/assets/26527928/24325156/b4af3922-119b-11e7-9d6e-1858c5689556.png)

Chronophylos commented:
You need to do
`git fetch upstream`
`git rebase upstream master`

Chronophylos commented:
Do you still want this to be done?

Chronophylos commented:
@ProstoSanja are you still working on this?

Chronophylos commented:
@chrisinajar Can you look over this? I want this to get merged as soon as possible.

Chronophylos commented:
Welp. OAA doesn't even launch anymore

warpdragon commented:
I can't believe you've done this
![image](https://cloud.githubusercontent.com/assets/13878439/24773518/94765f40-1ae3-11e7-9832-847b604fdac7.png)

Chronophylos commented:
What? I don't understand that picture..

Chronophylos commented:
ec1e975dc6f935b4f7b93b955d6b6efca7487420 and 0ba2478b7a0cf7f613b4788615122d35b3531614 crash too so the problem is not one of my commits

Chronophylos commented:
It's my Graphics driver supposedly.

Chronophylos commented:
I re-merged it because of even more conficts

chrisinajar commented:
NGP is currently disabled, and this is based on an ancient version of the codebase. When we revive ngp, we should come here for art assets and stuff, but unfortunately this is RIP
